### ISSUE TITLE
Support for HTTPS (TLS/SSL)

### ISSUE THREAD

**Issue by NoodleBB** [wekan/wekan#916]
When setting the `ROOT_URL` to `https` and try to access wekan via browser i just get 

> ERR_CONNECTION_CLOSED

. 

wekan v.0.13, v.0.12, v.0.10.1 _Manual installation on CentOS_

**Comment by davydov-vyacheslav** [comment]
one of the possible workaround - is to wrap wekan with apache and do smoething like that:
```

<VirtualHost *:443>

        SSLEngine on
        SSLCertificateFile /etc/apache2/ssl/some.cert
        SSLCertificateKeyFile /etc/apache2/ssl/some.key

        ServerName servername
        ServerAdmin webmaster@mytodo.org

        DocumentRoot /var/www/wekan.site.com
        <Directory />
                Options FollowSymLinks
                AllowOverride AuthConfig FileInfo Indexes Options=MultiViews
        </Directory>

        <Directory /var/www/wekan.site.com>
                Options -Indexes +FollowSymLinks +MultiViews
                AllowOverride AuthConfig FileInfo Indexes Options=MultiViews
                Require all granted
        </Directory>

        ErrorLog /var/log/apache2/wekan.site.com-error.log

        # Possible values include: debug, info, notice, warn, error, crit,
        # alert, emerg.
        LogLevel warn

        CustomLog /var/log/apache2/wekan.site.com-access.log combined
        ServerSignature Off

        ProxyPassMatch   "^/(sockjs\/.*\/websocket)$" "ws://127.0.0.1:8085/$1"
        ProxyPass        "/" "http://127.0.0.1:8085/"
        ProxyPassReverse "/" "http://127.0.0.1:8085/"

</VirtualHost>

```

where 8085 - is your local wekan's port

Don't know where ... but somewhere I saw such installation ..

**Comment by NoodleBB** [comment]
I don't have root access to that server.

**Comment by iAdanos** [comment]
@NoodleBB, did you set up HTTPS in webserver config correctly? 
You should attach your webserver config here.

If you trying to start "stock" wekan in HTTPS mode (doesn't matter docker instance or compiled from source), it will not work, because wekan needs webserver proxying it with SSL. 

Offtop: IMHO, SSL is not the function Wekan itself must provide. It's an app. Commutity must be concentrated on functions and stability of the primary service. 
I beleve, we want make a lightweight, fast and cool service. Making a monster with integrated webserver, DB etc. functions will kill the idea.

**Comment by NoodleBB** [comment]
```
RewriteEngine On
RewriteCond %{HTTPS} !=on
RewriteCond %{ENV:HTTPS} !=on
RewriteRule .* https://${SERVER_NAME}%{REQUEST_URI} [R=301,L]
RewriteRule ^(.*) http://localhost:61199/$1 [P]
```

**Comment by xet7** [comment]
Does Meteor / Node.js have support for virtual hosts, redirects, let's encrypt etc? URL?

**Comment by xet7** [comment]
That force-ssl seems to just redirect

**Comment by NoodleBB** [comment]
Yes. Im not an Node.js (or Meteor) expert. But Node has [TLS support](http://node.readthedocs.io/en/latest/api/tls/). 

How-Tos:
- https://docs.nodejitsu.com/articles/cryptography/how-to-use-the-tls-module/
- https://www.sitepoint.com/how-to-use-ssltls-with-node-js/
- http://www.giacomovacca.com/2015/02/websockets-over-nodejs-from-plain-to.html

**Comment by xet7** [comment]
Does somebody use only Node.js in production?

In all articles I found with Google search, Nginx/Caddy/some other webserver is used for load balancing and delivering static content in front of Node.js.

**Comment by iAdanos** [comment]
@xet7, i tried to tell this. Wekan is a backend application. User can have multiple wekan instances and a load balancer proxying them. It's a classic practice.
Using only backend application in production is a very bad idea. I think, nobody really do that. No caching, no access log, no filters, no load balancers, no access control.

The best example for me is a [Ghost Blogging Platform](https://ghost.org/). It can be run standalone, but developpers assume that it is always behind Nginx/Apache/whatever.
The application itself does not provide SSL, virtual hosts etc.
I think, that's the right way.

I use Caddy to provide Wekan HTTPS access, and i have almost no problems.

@NoodleBB doesn't have root access to server, that's his main problem. With all possible respect, it's not Wekan functionality problem.

**Comment by NoodleBB** [comment]
The majority don't have root access to the server they are using for hosting.

**Comment by REJack** [comment]
It's important which Control Panel your hoster/ISP use, if it's Plesk or CPanel you have not much problems depending on the restrictions.

I use Wekan on my own little root server with Deb 8, Plesk 17, a nginx proxy and Lets Encrypt SSL Certificate 😄 and it runs perfectly.

**Comment by NoodleBB** [comment]
I use SSH. I don't use a control panel.

**Comment by REJack** [comment]
@NoodleBB Which kind of access you have on your server trough SSH, do you can modify stuff in /etc, /var & /opt or install new packages?

**Comment by NoodleBB** [comment]
Nope.

**Comment by REJack** [comment]
Ok thats a bit hard. You need to check which modules are available in the apache2 instance. 
One method to check which modules are available, create empty phpfile and add `<?php phpinfo() ?>` then search for apache2handler. 

If there are proxy modules avaible you can setup a reserve proxy like @davydov-vyacheslav's workaround.

These lines can be added into a .htaccess file too.
```
        ProxyPassMatch   "^/(sockjs\/.*\/websocket)$" "ws://127.0.0.1:61199/$1"
        ProxyPass        "/" "http://127.0.0.1:61199/"
        ProxyPassReverse "/" "http://127.0.0.1:61199/"
```

**Comment by NoodleBB** [comment]
@REJack 
I'll take a look tomorrow.

@xet7 
Sure. Not in global or enterprise setups, but in infinity small and private groups like families, school classes, fraternities, walking club, dance groups, boycouts, churches, choral society, neighborhood, etc., etc.

**Comment by NoodleBB** [comment]
@xet7 
Sure. Not in global or enterprise setups, but in infinity small and private groups like families, school classes, fraternities, walking club, dance groups, boycouts, churches, choral society, neighborhood, etc., etc.

**Comment by Zokormazo** [comment]
I prefer using reverse proxy with nginx instead of exposing the nodejs socket to internet so https goes on nginx. Better practice

**Comment by itsmedani** [comment]
I all,

I tried to configure Apache for SSL connection. I have SSL connections until I opened a card. 

Indeed, the URL switch to the URL which are configured on the URL_ROOT line. 

Do you know why?

Regards

**Comment by NoodleBB** [comment]
Try to set the protocol of the URL_ROOT to **https://**

**Comment by itsmedani** [comment]
Hi,

It's work!

Thanks :)

**Comment by bitsandnumbers** [comment]
@Zokormazo and what if you use docker ? the dockerfile wekan team provide does not include a web server (nginx, apache) other than node.js, so reverse-proxy with https won't work : no way to set ssl keys (setting them in nginx proxy host is not enough it seems). Or am I missing something ?

**Comment by Zokormazo** [comment]
why not? I'm using official docker image behind nginx reverse proxy with
let's encrypt certificates.

2017 eka. 13 15:35 erabiltzaileak hau idatzi du ("bitsandnumbers" <
notifications@github.com>):

**Comment by bitsandnumbers** [comment]
Yeah, sorry I deleted my comment after realizing that it could work. I actually found my solution.

**Comment by supercrypt** [comment]
Hey all,

we are using Wekan with Stunnel4 to support encrypted connections over SSL and want to activate HSTS. Would it possible to integrate this in a future release?

As far I can see it's already supported by Node:
https://stackoverflow.com/a/29293551

**Comment by xet7** [comment]
@supercrypt

Does that http-proxy include support for SSL ?

Otherwise, if it requires separate install of stunnel, it's the same as using Nginx/Caddy etc in front of Wekan.

**Comment by xet7** [comment]
@supercrypt 

Is stunnel included inside http-proxy npm package? So there is no need for stunnel setup separately?

**Comment by supercrypt** [comment]
@xet7 I don't know if this is the case.

**Comment by cknoll** [comment]
@NoodleBB As you are the main author of <https://github.com/wekan/wekan/wiki/Install-latest-Wekan-release-on-Uberspace>, I guess this issue is related to that platform. Have you in the meantime been able to server Wekan on uberspace with https?

**Comment by y0d4a** [comment]
is there news about this feature?

**Comment by patrickdung** [comment]
Securing the http port with TLS is important.
If the reverse proxy web server connect to wekan's/nodejs http port without encryption, the data communication between them can be sniffed.
stunnel can be setup to secure the proxy traffic but it would be less convenient.

**Comment by xet7** [comment]
@patrickdung 

Having Nginx/Apache/Caddy/stunnel to provide SSL in front of Wekan is already possible.

This issue is only about having SSL in Wekan's node.js server itself.

**Comment by y0d4a** [comment]
little more help? :)
